tools/lister.py

```diff
@@ -0,0 +1,97 @@
+#!/usr/bin/env python
+from __future__ import print_function
+
+import os
+import sys
+import subprocess
+import re
+from collections import defaultdict
+
+def get_ftype(fpath, use_shebang):
+    ext = os.path.splitext(fpath)[1]
+    if ext:
+        return ext[1:]
+    elif use_shebang:
+        # opening a file may throw an OSError
+        with open(fpath) as f:
+            first_line = f.readline()
+            if re.search(r'^#!.*\bpython', first_line):
+                return 'py'
+            elif re.search(r'^#!.*sh', first_line):
+                return 'sh'
+            elif re.search(r'^#!.*\bperl', first_line):
+                return 'pl'
+            elif re.search(r'^#!', first_line):
+                print('Error: Unknown shebang in file "%s":\n%s' % (fpath, first_line), file=sys.stderr)
+                return ''
+            else:
+                return ''
+    else:
+        return ''
+
+def list_files(targets=None, ftypes=None, use_shebang=True, modified_only=False,
+    exclude=None, group_by_ftype=False):
+    """
+    List all files tracked by git.
+    If targets is not None, only those files will be returned which are either
+    in targets or they are in a directory in targets.
+
+    Arguments:
+
+    targets - List of paths to directories where search is to be performed.
+        None searches current directory, [] returns [] or {}.
+    ftypes - List of file types on which to filter the search.
+        None includes all file types, [] returns [] or {}.
+    use_shebang - Determine file type of extensionless files from their shebang.
+    modified-only - Only include files which have been modified.
+    exclude - List of paths to be excluded. None acts like [].
+    group_by_ftype - If True, returns a dict of lists keyed by file type.
+        If False, returns a flat list of files.
+    """
+    if targets==[] or ftypes==[]:
+        return []
+    targets = targets or []
+    exclude = exclude or []
```

Comment:
Why not just use `[]` as the default in the function definition?  I can think of a advantages that has, but I'm curious what your reasoning is and how you trade that off against the extra lines of code.